### PR TITLE
Propagate exceptions raised in Python callback functions

### PR DESCRIPTION
**Pull request by madsbk**
## Description
Currently, we ignore Python exceptions raised in callback functions called by resources such as `CallbackMemoryResource`. This is because Cython doesn't automatically translate Python exceptions into C++ exceptions (even though Cython does translate C++ exceptions into Python automatically).

To address this issue, we now translate Python exceptions raised in the callback function into a C++ exception manually. Next, we remove Cython's [automatic C++ to Python exception translation](https://cython.readthedocs.io/en/stable/src/userguide/wrapping_CPlusPlus.html#exceptions) by **not** using the `except +` declaration. Thus, when the callback raises an exception, the caller (a memory resource implemented in C++) can catch and/or propagate the exception.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/rmm/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: python/rmm/_lib/memory_resource.pyx [hunk 1]
```diff
@@ -523,8 +560,14 @@ cdef void* _allocate_callback_wrapper(
...
-) with gil:
-    return <void*><uintptr_t>((<object>ctx)(nbytes))
+) nogil:
+    cdef CppExcept err
+    with gil:
+        try:
+            return <void*><uintptr_t>((<object>ctx)(nbytes))
+        except BaseException as e:
+            err = translate_python_except_to_cpp(e)
+    throw_cpp_except(err)
```
Comment (shwina) [inline python/rmm/_lib/memory_resource.pyx:570]: I don't quite understand -- won't the C++ exception thrown here end up _not_ propagating to Python (this function is not marked `except +`)?

File: python/rmm/_lib/memory_resource.pyx [hunk 2]
```diff
@@ -35,6 +36,42 @@ from rmm._cuda.gpu import (
...
+    propagate exceptions.
+    """
+    if isinstance(err, MemoryError):
+        return CppExcept(0, str.encode(str(err)))
+    if isinstance(err, BaseException):
+        return CppExcept(-1, str.encode(str(err)))
+
+# Implementation of `throw_cpp_except()`, which throws a given `CppExcept`.
+# This function MUST be called without the GIL otherwise the thrown C++
+# exception are tanslated back into a Python exception.
```
Comment (bdice) [inline python/rmm/_lib/memory_resource.pyx:58]:
```suggestion
# exception are translated back into a Python exception.
```

File: python/rmm/_lib/memory_resource.pyx (hunk 1 above)
Comment (madsbk) [reply python/rmm/_lib/memory_resource.pyx:570]: Yes, thus the caller (a memory resource implemented in C++) can now catch and/or propagate the exception.

Comment (shwina) [reply python/rmm/_lib/memory_resource.pyx:570]: Ah I see, that makes sense. Thanks!

**Review by shwina — APPROVED** [review]

**Comment by madsbk** [comment]
@bdice anything else needed here?

File: python/rmm/_lib/memory_resource.pyx [hunk 3]
```diff
@@ -35,6 +36,42 @@ from rmm._cuda.gpu import (
...
+    The returned exception handle can then be thrown by `throw_cpp_except()`,
+    which MUST be done without holding the GIL.
+
+    This is useful when C++ calls a Python function and needs to catch or
+    propagate exceptions.
+    """
+    if isinstance(err, MemoryError):
+        return CppExcept(0, str.encode(str(err)))
+    if isinstance(err, BaseException):
+        return CppExcept(-1, str.encode(str(err)))
```
Comment (vyasr) [inline python/rmm/_lib/memory_resource.pyx:54]: Let's make it clear that this is a base case, all exceptions subclass from `BaseException`.
```suggestion
    return CppExcept(-1, str.encode(str(err)))
```

**Review by vyasr — COMMENTED** [review]
One suggestion for improvement here, then I can approve. Bradley is out for most of this week.

File: python/rmm/_lib/memory_resource.pyx (hunk 3 above)
Comment (madsbk) [reply python/rmm/_lib/memory_resource.pyx]: Good point!

**Review by vyasr — APPROVED** [review]

**Comment by vyasr** [comment]
@gpucibot merge

**Pull request merged**